legal-api/src/legal_api/models/db.py

```diff
@@ -11,17 +11,200 @@
+
+    @classmethod
+    def _check_versioning(cls):
+        from legal_api.services import flags
+        current_service = current_app.config['SERVICE_NAME']
+        db_versioning = flags.value('db-versioning')
+        use_new_versioning = (bool(db_versioning.get('initialize-new-versioning'))
+                            and bool(db_versioning.get('enable-new-versioning', {}).get(current_service)))
+        cls._current_versioning ='new' if use_new_versioning else 'old'
+        print(f'\033[31mCurrent versioning={cls._current_versioning}\033[0m')
+        
+    @classmethod
+    def _initialize_versioning(cls):
+        cls._is_initialized = True
+        cls._check_versioning()
+        disabled = 'new' if cls._current_versioning == 'old' else 'old'
+        cls._versioning_control[disabled]['disable']()
+
+    @classmethod
+    def _switch_versioning(cls, previous, current):
+        cls._versioning_control[previous]['disable']()
+        cls._versioning_control[current]['enable']()
+
+    @classmethod
+    @debug
+    def unlock_versioning(cls, session, transaction):
+        print(f'Session={session}, transaction={transaction}')
+        if '_versioning_locked' in session.info and '_transactions_locked' in session.info:
+            session.info['_transactions_locked'].remove(transaction)
+            print('\033[35mTransaction unlocked\033[0m')
+
+            if not session.info['_transactions_locked']:
+                session.info.pop('_versioning_locked', None)
+                session.info.pop('_transactions_locked', None)
+                print('\033[31mVersioning unlocked\033[0m')
+            
+            # debug - else statement
+            else:
+                print(f"This session has active transaction, can't be unlocked")
+
+        # debug - else statement
+        else:
+            print("Versioning/Transaction lock doesn't exist, skip")
+
+    @classmethod
+    @debug
+    def get_transaction_id(cls, session):
+        transaction_id = None
+        current_versioning = session.info['_versioning_locked']
+
+        print(f'\033[31mCurrent versioning type={current_versioning}\033[0m')
+        transaction_id = cls._versioning_control[current_versioning]['get_transaction_id'](session)
+        print(f'\033[31mUsing transaction_id = {transaction_id}\033[0m')
+
+        return transaction_id
+
+    @classmethod
+    @debug
+    def version_class(cls, session, obj):
+        if not session.in_transaction():  # trigger versioning setup listener
+            session.begin()
+
+        current_versioning = session.info['_versioning_locked']
+        print(f'\033[31mCurrent versioning type={current_versioning}\033[0m')
+
+        return cls._versioning_control[current_versioning]['version_class'](obj)
+
+
+@debug
+def setup_versioning():
+    @event.listens_for(SignallingSession, 'after_transaction_create')  # skip events for continuum's internal session/txn operations
```

Comment:
`SignallingSession` is the default session that our flask app uses and it extends the core session. I use it to distinguish the session that `continuum` uses and want to make it easier to track transaction lifecycle. `continuum` uses core session for its internal operations. It creates transaction in its own session and detaches it from its session and add it to the flask session.